### DISCUSSION TITLE
Little IMU demo

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#1150] [Show and tell]
I'm working on an attitude sensor that tracks the roll & pitch of an ocean racing kayak (surfski) and in order to get a sense of the data quality I put together a little graphics demo using a moddable two. It's now showing signs of life :-).

Stuff used:

- moddable two (esp32)
- Bosch BMI160 accelerometer & gyroscope
- Xio Tech sensor fusion (https://github.com/xioTechnologies/Fusion)
- Poco outline drawing

![image](https://github.com/Moddable-OpenSource/moddable/assets/39480/16c2f469-d8ee-48b5-88d8-a9366003945f)

Code architecture:
- worker runs at 100Hz to fetch IMU data
- at 16Hz worker pipes IMU data through sensor fusion algorithm and outputs position data (roll/pitch/yaw)
- main thread consumes position data, does 3D projection, and draws rendering of PCB

Short demo video: https://tve.s3.amazonaws.com/public/imu-demo-20230610-1300.mp4

Getting 16fps rendering is pretty darn amazing, Poco rocks!

**Comment by mkellner** [comment]
Alas, the .mp4 link leads me to this:
```
<Error>
<Code>PermanentRedirect</Code>
<Message>The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.</Message>
<Endpoint>tve.s3.amazonaws.com</Endpoint>
<Bucket>tve</Bucket>
...
```

**Reply by tve** [reply]
It's my fat-finger-everything day 😢
Now fixed

**Comment by tve** [comment]
Source code is available at:
- bmi160 driver: https://github.com/tve/modtve/tree/main/modules/drivers/sensors/bmi160
- bmi160viz app: https://github.com/tve/modtve/tree/main/modules/examples/bmi160viz
- sensor fusion: https://github.com/tve/Fusion

All these are work-in-progress!

**Reply by phoddie** [reply]
Very cool!! Thanks for sharing the your code. The Sensor Fusion library port is going to be very useful.

Nice use of Poco outlines. The performance is really impressive - it feels entirely responsive in real-time. I see you are taking advantage of the partial update feature, which really makes a difference. I also notice that you bumped the size of the default `pixels` by 4x to render larger slabs. That helps to reduce computation overhead when rendering, especially with outlines.

In reading through your code, I have a few observations - some about performance and others just about general correctness. If you are open to it, I can write those up for you.

**Reply by tve** [reply]
Thanks for the encouraging comment! Yes, I would very much love feedback of any kind and appreciate the time it takes to write that up.

On the performance front doing the 4x4 matrix multiply in JS is slow. Stuff I'm less clear about is how accessing local vars, vs. instance props, vs. module vars compares.

On the display front, it's really blow-me-away nice: totally real-time. At ~15 frames/sec the main limitation actually becomes flickering because the updates overlap with screen refreshes. I believe that given the moddable two HW it's not possible to sync.

**Reply by phoddie** [reply]
> Thanks for the encouraging comment! Yes, I would very much love feedback of any kind and appreciate the time it takes to write that up.

Cool. Will do.

> On the performance front doing the 4x4 matrix multiply in JS is slow. Stuff I'm less clear about is how accessing local vars, vs. instance props, vs. module vars compares.

You can optimize the matrix for access native code. Instead of storing the matrix as JavaScript properties, store them as a native data structure (Float64Array could be enough, since it is a buffer at the end of the day - use [xsmcGetBufferReadable](https://blog.moddable.com/blog/buffers/)). Then you can get a pointer to the buffer in C with no additional overhead to access the elements.

> On the display front, it's really blow-me-away nice: totally real-time. At ~15 frames/sec the main limitation actually becomes flickering because the updates overlap with screen refreshes. I believe that given the moddable two HW it's not possible to sync

Yes, synchronization is its own challenge. The display module we use on Moddable Two doesn't expose that pin, so it isn't possible there.

**Reply by phoddie** [reply]
Here are the notes I made on a pass through the code. I understand that this is very much a work-in-progress, so some of these points may already be things you are aware of. Looks very reasonable overall. The performance is all-the-more impressive after reading through the various layers of code to see everything that's going on.

## BMI160 driver

Nice to see this implemented as an ECMA-419 Sensor class. It should work with Node-RED MCU Edition's Sensor node. Once the dust settles, it would be good to document the details into [TR/109](https://www.ecma-international.org/wp-content/uploads/ECMA_TR-109_1st_edition_june_2021.pdf) ["Class specifications for embedded hardware components"](https://github.com/EcmaTC53/spec/tree/master/docs/Hardware%20Components/sensors).

- [This call](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/drivers/sensors/bmi160/bmi160.ts#L316) to `Object.freeze` is unnecessary. The XS linker has done this automatically for some time now. It hasn't been cleared out of all the implementations, but there's no need for it in new code.

- If [these reads](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/drivers/sensors/bmi160/bmi160.ts#L108-L116) throw an exception (e.g. the sensor isn't present), the this.#io will be orphaned. That's usually not the end of the world. In more complete implementations, we'll use a try/catch block and expressly close io if something goes wrong.

- To strictly following ECMA-419, [`configure` is required](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/drivers/sensors/bmi160/bmi160.ts#L154-L172). It can just be a stub.

- If `sample()` doesn't have a sample to return, it should just return `undefined`. `onError` should only be called on a fatal error. There's no standard mechanism to report why the sample isn't available.

- This is returning a reference to the private buffer `#regsRaw`. Those get overwritten in places, notably `sample()`. Maybe you intend to make a copy here?

- By definition, calling `close` more than once should not throw so [this](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/drivers/sensors/bmi160/bmi160.ts#L177) should be `this.#io?.close()`


## Fusion module

- For performance, use the xsID_* constants in place of [`xsID("x")`](https://github.com/tve/Fusion/blob/983f80ea12a66e826d26e3a6edb7994047d63de4/fusion.c#L39). You need to `#include "mc.xs.h"`.  It isn't required but it is smaller and faster. Doing this thoroughly would required adding a parameter to `defineProperty` for the ID value in addition to the key name.

- `xsmcVars` is incremental on the current stack frame, so each time it is called it adds another variable. Since [`extractVector` ](https://github.com/tve/Fusion/blob/983f80ea12a66e826d26e3a6edb7994047d63de4/fusion.c#L38) is called three times from `xs_FusionAhrsUpdate` it will allocate more than needed. The usual approach is to call xsmcVars in `xs_FusionAhrsUpdate` with what is required by the functions it calls. (Similar note for `defineProperty`.

- In `xs_FusionAhrsGetQuaternion` you can use `xsmcSetIndex` instead of `xsmcSetAt` to save setting up one slot.

## bmi160viz app

- Looking at `multiply()` and friends in the Matrix4 class, you are set-up nicely to move the heavy path to C. Since you have the matrix data already in a `Float64Array` it should be very straightforward. It will be faster.

- In imu.ts, [these two lines](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/examples/bmi160viz/imu.ts#L33-L34) could be more efficient and compact using `Uint8Array.of`. Given that it is outside a loop, it won't matter for performance. But, I noticed it so thought I'd pass it along.

- Nice use of `Math.idiv()` [here](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/examples/bmi160viz/imu.ts#L40). Reminds me that I should do the work to get that to Stage 2.... and add it to `xs.d.ts` so you don't have to declare it here.

- Tiny detail - `Timer.clear()` ignores the argument if it is not an object, so you can remove the "if" [here](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/examples/bmi160viz/imu.ts#L84).

- Is [this delay](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/examples/bmi160viz/imu.ts#L26) needed? It seems like the sensor driver already takes care of any necessary waiting.

- In imuWorker, this is an interesting way to get the [device global](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/examples/bmi160viz/imuworker.ts#L7). I'm a little surprised it works! The `device` global should just be a shortcut to access the [host provider instance](https://419.ecma-international.org/#-16-host-provider-instance) at `embedded:provider/builtin`. Maybe give that a try?

- Small point, but in main.ts since this is an inner-ish loop, it would be worth creating a local variable for `faces[i]` [here](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/examples/bmi160viz/main.ts#L126-L128) instead of performing the look-up each time.

- A constructor should never return `undefined`. If it fails, it should throw. That's definitely the case with Worker as confirmed by [our tests](https://github.com/Moddable-OpenSource/moddable/blob/public/tests/modules/base/worker/constructor.js). So, [this check on the Worker constructor](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/examples/bmi160viz/main.ts#L264-L265) should be unnecessary.

- JavaScript execution in a VM is single threaded, even with workers. As a result, `updateQ` cannot change asynchronously [in this loop](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/examples/bmi160viz/main.ts#L274-L280). So, I **think** you can simplify the [`onmessage` function](https://github.com/tve/modtve/blob/0c6129a1576ccd3d9b201f7afad277946e574207/modules/examples/bmi160viz/main.ts#L269-L287):

```js
  imu.onmessage = function (message: any) {
      updateQ = message
      updateTimer ?= Timer.set(() => {
        updateDisplay(updateQ)
        updateQ = null
        updateTimer = null
      }, 1)
    }
  }
```

- That timer is used to trigger frame drawing, so that happens many times a second, allocating and freeing a timer for each frame. You can allocate the timer once and schedule it as needed to eliminate some overhead. A convenient way to do that is to create it and immediately unschedule it

```js
let timer = Timer.repeat(() => {/* code here */}, 100);
Timer.schedule(timer);		// unschedule
```

Then you can schedule it when needed using the full form of `Timer.schedule()`. Since you are doing one-shots, be sure to set the repeat to a non-zero value or it will behave like `set` and clear itself after firing.

```js
Timer.schedule(timer, 1, 100_000);
```
_ In reading this, I think there are opportunities to reduce the creation of new objects which would reduce the GC overhead. I'm not really sure without running the code though to see what's going on, so I'm not making any suggestions. Just noting the potential.

- Rather than set a sample rate to the IMU worker, maybe the main thread should send a message to indicate it is ready for the next reading. It could do that **before** drawing the current frame, so that the IMU worker can prepare that in parallel and have it waiting after the current frame finished drawing. There are probably other approaches, but the idea would be to avoid hard coding the sampling frequency but instead adapt to the rendering rate.

**Reply by tve** [reply]
Phew, all great comments, thanks a ton!!! It really helps to have someone who has decades of experience in a language look things over.

There is one comment I don't understand, which is getting the `device` global in a worker. When I looked in xsbug I did not see that global in the worker. Are you saying that I should `import { device } from "embedded:provider/builtin"`? I tried passing the device in a message and to my surprise that worked so I didn't look any further.

WRT sample rate, etc. what I did was to add a batch mode to the driver. The IMU has a buffer where it can store 83 samples, which is enough for 0.83 seconds @100Hz. So the latest version puts the driver in that mode and then just calls `bmi160.batch(sample => { ... })` in the main loop. This completely eliminates the worker and everything is much simpler. Only downside is that there's no overlap of IMU/AHRS and display possible. The main loop is then:
```
  startIMU()
  let run = true
  function loop() {
    const q = sampleIMU() // calls bmi160.batch() and processes the samples, returning the final attitude quaternion
    if (q) updateDisplay(q)
    if (run) Timer.set(loop, 0)
  }
  Timer.set(loop, 0)
```
You point about reallocating a Timer is still pertinent... Dunno whether there's a more canonical way to write a near-infinite loop which keeps tasks flowing so instrumentation works?

**Reply by tve** [reply]
I pushed the updated version of the bmi160 driver that has a [batch() function](https://github.com/tve/modtve/blob/main/modules/drivers/sensors/bmi160/bmi160.ts#L252-L283). This function sits in a loop where it reads out the samples from the device's fifo and calls a callback function with each sample in turn.

This works really well and simplifies the overall app design a lot: no more worker and no more 100Hz timer. But I'm not thrilled by having batch() and sample() to retrieve samples. I think I can do another iteration and merge the two, hiding the fifo inside of sample() and eliminating the callback. The use_fifo=true/false toggle would then only affect which samples are returned by the function: if false then it's always the latest, and if true it's from the fifo, i.e., potentially older. To empty the fifo, one would do something like:
```
const t0 = Time.ticks
let sample = bmi160.sample()
while (sample) {
  process_sample(sample)
  if (sample.ticks > t0) break // got all samples queued when we started
  sample = bmi160.sample()
}
```

**Reply by phoddie** [reply]
I'll miss the Worker. The ESP32 has two cores and it is great to use both when it helps. But, it is a bit more complicated so if it isn't necessary, you can hold that knowledge in reserve until it is needed.

For your timer you can take advantage of the fact that the repeating timer is drift free. If you schedule a repeating timer with an interval of 1 ms but do more than 1 ms of work, it will give the behavior you want -- a non-blocking event loop that calls you back without delay:

```js
Timer.repeat(()=> {
}, 1);
```

I like the option you propose to have the sensor driver return either the latest reading or the oldest in the FIFO. Does the sensor FIFO have a time stamp of some sort so you can return an accurate time stamp for the readings?

> There is one comment I don't understand, which is getting the device global in a worker. When I looked in xsbug I did not see that global in the worker. Are you saying that I should import { device } from "embedded:provider/builtin"? I tried passing the device in a message and to my surprise that worked so I didn't look any further.

Basically... but `import device from "embedded:provider/builtin"` as in the ECMA-419 provided above for "host provider instance."

Passing the device global to the worker succeeds because XS is able to marshall preloaded functions across the worker boundary. That would never work in a browser. But. there's no need for it here because you can just access `device` from the worker through its module specifier.

**Reply by tve** [reply]
> Does the sensor FIFO have a time stamp of some sort so you can return an accurate time stamp for the readings?

Complicated... The chip can provide timestamps but I'm finding them useless here. The batch() function calculates and provides them in each sample. The time-stamps are based on the fact that the samples in the fifo are collected at a fixed interval and the assumption that if the fifo is not full the first sample was inserted one interval after the last sample previously retrieved. It's a bit inexact, not least because the chip's clock can drift quite a bit (+/-1%).

**Reply by phoddie** [reply]
Time is never easy. And the Bosch sensors often require some black magic to get everything working. I ask because time stamping of sample data is a popular topic in TC53. You can find the results of that in the timing notes in the standard and [TR/110](https://www.ecma-international.org/wp-content/uploads/ECMA_TR-110_1st_edition_june_2021.pdf). Not critical for your uses, but interesting to try to get right for the more general case.

**Reply by tve** [reply]
Yet another thought... One of the pieces in the demo code is the imu.ts module. It combines the raw bmi160 samples and the fusion library to provide processed attitude information. There's no reason this IMU module couldn't use the ECMA-419 Sensor pattern itself!
I have a similar situation with a power management chip (INA233) for which I'm writing a Sensor driver and then a "battery" module that keeps track of battery energy consumption. That battery module could also be written using the Sensor pattern.
I suppose the imu and battery modules wouldn't be called drivers, but that's OK and by following the pattern they would be easier to consume by others, I hope. Any thoughts about this approach?

**Reply by phoddie** [reply]
I like the idea. If the API is a comfortable enough fit, it makes sense to re-use.

**Comment by phoddie** [comment]
We bought a BMI160 sensor and set this up at the Moddable office today. Good news. It works! 🎉It feels just as magical and responsive to use at it looks in your video.

![tve](https://github.com/Moddable-OpenSource/moddable/assets/6856458/498a3341-2a60-42fe-abc7-020e9e127c12)

A few notes:

- The bold font mentioned in the manifest isn't in your repository. I substituted another.
- There's a module (meminfo?) included in the manifest that isn't in the repository. That seems unnecessary so I deleted the line.
- There's some flickering when drawing. That seems to be because of GC of outlines -- especially noticeable on the axes (you reported that separately). I added an array to keep them from being GC and the drawing is stable.
- I bumped the rendering buffer (Poco pixels) and got the frames rate to about 18 FPS.

**Reply by tve** [reply]
Yay! Thanks for the notes, I'll fix & good news about the flickering!

BTW: the camera angle was chosen to approximate the perspective in the video (which is also the one I had while developing). If you tack the device onto the moddable-two you may want to change the way things are represented...

**Reply by phoddie** [reply]
> If you tack the device onto the moddable-two you may want to change the way things are represented

Thanks. We did notice. It is awkward to use connected to the device. We're adding a ribbon cable.

Forgot to mention, you seem to be using a version of xs.d.ts that defines `Math.idiv()`. Without that, I got a build failure. I've updated that with all the integer math proposed functions.

I have an idea for how to fix the GC problem that should be general and efficient. Just need to find some time to implement it.

**Reply by tve** [reply]
Someone didn't do any release engineering... :-)
I have a pile of changes in my moddable tree that I need to review & commit & submit.. The idiv one is:
```
diff --git a/typings/global.d.ts b/typings/global.d.ts
index be64bba4..3405cf2f 100644
--- a/typings/global.d.ts
+++ b/typings/global.d.ts
@@ -6,4 +6,14 @@ declare global {
   // NOTE: `System` is non-standard and temporary to support the IO examples. Breaking changes are possible.
   var System: typeof System_;
   var device: typeof Device;
+
+  interface Math {
+    idiv(a: number, b: number): number
+    imod(a: number, b: number): number // (x%y+y)%y (always pos results)
+    irem(a: number, b: number): number // x%y (neg x -> neg result)
+    idivmod(a: number, b: number): [number, number] // [x/y, (x%y+y)%y]
+    imuldiv(a: number, b: number, c: number): number // (a * b) / c
+    mod(a: number, b: number): number // (x%y+y)%y (floating point)
+  }
+
 }
```

**Reply by phoddie** [reply]
We do accept PRs. 😉

**Comment by tve** [comment]
I updated the BMI160 driver, broke out an AHRS library, tweaked the Fusion library, and updated the main app:
- the worker is back! there's an `if` statement at the end of main.js to toggle between no-worker and worker
- almost all of phoddie's comments are integrated (thanks again!)
- the matrix math is still in JS
- the AHRS library is independent of the BMI160 driver
- it now runs at 20 FPS 🚀🚀

The worker performs the AHRS stuff, the main does the 3D projection and the drawing. One CPU is 98% busy the other 22%. It would probably be yet faster if the 3D calcs were moved to the worker. But I fear that's beyond my patience now 😆.

I patterned the AHRS library after the Sensor class pattern but I don't like it. As it stands the information to construct the IMU device is passed in ECMA-419 style and the AHRS class constructs the device. But all it really does with it is to call `sample()`. So I want to do another round and remove the IMU stuff from AHRS and let it have an `update()` method into which the sample is passed. All this really changes is that the caller of AHRS instantiates the IMU and instead of `ahrs.sample()` it does `ahrs.update(imu.sample())`. Two advantages are that the caller holds the IMU object and can perform other operations on it easily (for example, sleep and awaken on motion), and the caller can transform the sample object if the IMU driver doesn't produce exactly the right thing.

(I didn't do a fresh install based on what's in the repo, please let me know if I forgot something again...)

**Reply by phoddie** [reply]
Sounds great! 

I temporarily lost my BMI160 sensor hardware, so I can't try it at the moment. I'll have it back in a day or two.

> the worker is back! there's an if statement at the end of main.js to toggle between no-worker and worker

Very nice to have both versions available to experiment.

> One CPU is 98% busy the other 22%.

Yea, when rendering something non-trivial flat-out, one core ends up pretty busy.

> the matrix math is still in JS

Is that mostly `Matrix4.multiply()`? Can I help with that?

**Reply by tve** [reply]
Haha, you should have bought the 3-pack! 😂
Here's a profile snapshot of the main thread:
![image](https://github.com/Moddable-OpenSource/moddable/assets/39480/c5469edb-34de-4bc4-92ea-a3e66fc71a61)
Which features Matrix4 operations prominently... That's all in `hack3d.js`, which is a bunch of functions someone wrote more or less from textbook to see a bottom-up implementation of that stuff himself. I copied it 'cause I was looking for a quick and small solution and that fit the bill. The Matrix4 class should be reusable so it's perhaps worth breaking it out and working on it. The rest I'm not so sure... The whole thing is a little test app that is threatening to get out of hand, LOL!
Oh, also, back face elimination (face normal vector pointing away from the camera) would cut the number of polygons in half... Then you could get the same frame rate with a tad more interesting model...
But to your question: please have fun with it! PRs happily accepted but if it's easier to copy & reorg then enjoy!

BTW: what do the two numbers in the profile mean? self and total-with-children? Is that the same in top-down vs. bottom up mode? Might be nice to write a section in the xsbug.md?

**Reply by mkellner** [reply]
@tve, I don't know if you saw these blog posts, but they have some detail on the profiler.

https://blog.moddable.com/blog/profiler/

https://blog.moddable.com/blog/optimizing-life/

**Reply by tve** [reply]
I did see them but I didn't see any description of what the numbers mean...

**Reply by phoddie** [reply]
I've never met a profiler that didn't take me some time to learn to interpret the results. The XS profiler follows the approach of other JavaScript profilers to try to minimize the learning curve (we tested by exporting XS profile data to view other tools). A simple test program often helps me understand the profiler output.

```js
import Timer from "timer"

function leaf() {
	for (let i = 0; i < 100; i++)
		i * i * i;
}

function runner() {
	for (let j = 0; j < 100; j++)
		leaf();
}

Timer.repeat(runner, 1);
```

Running that in the simulator gives me this:

<img width="437" alt="image" src="https://github.com/Moddable-OpenSource/moddable/assets/6856458/141ec128-7d32-41d6-a1cc-71853fe905d6">

That makes it clear that the left column is time in the function itself and the right column is elapsed time in the function (function + children).

Note that XS implements a sampling profiler, so the results are likely not identical across runs and can miss functions that don't execute for long and are called infrequently. As a JavaScript profiler, it doesn't have visibility into the native host code. All that said, it is incredibly useful for quickly identifying hotspots and evaluating the impact of optimizations.

**Reply by phoddie** [reply]
Detailed documentation on the XS Profiler implementation is in this repository in [XS Profiler.md](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/XS%20Profiler.md). The [Viewers section](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/XS%20Profiler.md#viewers) explains how the numbers displayed are calculated.

**Comment by phoddie** [comment]
I had some time to rework the `Matrix4` class to use C in a few places. The API should be unchanged. It still works, at least. ;) It pretty much eliminates the Matrix4 calls from any prominent place in the profiles, which seems good. I made the constructor native as well, as it eliminates some work that happens when using `...args`.

Here are the sources:

<details>
<summary>hack3d.d.ts</summary>

```ts
// Based on https://github.com/jordoreed/home-grown-graphics/blob/main/index.html
// By Jordan Reed

declare module "hack3d" {

  function randomFloat(min: number, max: number): number
  function toRadians(degrees: number): number

  interface Vec4 {
    x: number
    y: number
    z: number
    w: number
  }

  class Matrix4 {
    m: Float64Array

    constructor(...args: number[])
    constructor(arr: Float64Array)
    constructor(...args: number[] | Float64Array[])

    multiply(r: Matrix4): Matrix4
    multiplyVec4(v: Vec4): Vec4

    static identity(): Matrix4

    static perspective(l: number, r: number, b: number, t: number, n: number, f: number): Matrix4

    static perspectiveAspectRatio(aspectRatio: number, fov: number, n: number, f: number): Matrix4

    static translate(x: number, y: number, z: number): Matrix4

    static scale(x: number, y: number, z: number): Matrix4

    // pitch
    static rotateX(degrees: number): Matrix4

    // yaw
    static rotateY(degrees: number): Matrix4

    // roll
    static rotateZ(degrees: number): Matrix4
  }
}
```

</details>

<details>
<summary>hack3d.js</summary>

```js
// Based on https://github.com/jordoreed/home-grown-graphics/blob/main/index.html
// By Jordan Reed

export function randomFloat(min, max) {
  return Math.random() * (max - min) + min
}

export function toRadians(degrees) {
  return (degrees * Math.PI) / 180
}

export class Matrix4 {
  constructor() @ "xs_Matrix"
  multiply(r) @ "xs_matrix_multiply"
  multiplyVec4(v) @ "xs_matrix_multiplyVec4"

  static identity() {
    return new Matrix4(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1)
  }

  static perspective(l, r, b, t, n, f) {
    return new Matrix4(
      (2 * n) / (r - l),
      0,
      (r + l) / (r - l),
      0,
      0,
      (2 * n) / (t - b),
      (t + b) / (t - b),
      0,
      0,
      0,
      -(f + n) / (f - n),
      (-2 * f * n) / (f - n),
      0,
      0,
      -1,
      0
    )
  }

  static perspectiveAspectRatio(aspectRatio, fov, n, f) {
    const halfH = Math.tan(toRadians(fov * 0.5)) * n
    const halfW = aspectRatio * halfH
    return Matrix4.perspective(-halfW, halfW, -halfH, halfH, n, f)
  }

  static translate(x, y, z) {
    return new Matrix4(1, 0, 0, x, 0, 1, 0, y, 0, 0, 1, z, 0, 0, 0, 1)
  }

  static scale(x, y, z) {
    return new Matrix4(x, 0, 0, 0, 0, y, 0, 0, 0, 0, z, 0, 0, 0, 0, 1)
  }

  // pitch
  static rotateX(degrees) {
    const radians = toRadians(degrees)
    const c = Math.cos(radians)
    const s = Math.sin(radians)
    return new Matrix4(1, 0, 0, 0, 0, c, -s, 0, 0, s, c, 0, 0, 0, 0, 1)
  }

  // yaw
  static rotateY(degrees) {
    const radians = toRadians(degrees)
    const c = Math.cos(radians)
    const s = Math.sin(radians)
    return new Matrix4(c, 0, s, 0, 0, 1, 0, 0, -s, 0, c, 0, 0, 0, 0, 1)
  }

  // roll
  static rotateZ(degrees) {
    const radians = toRadians(degrees)
    const c = Math.cos(radians)
    const s = Math.sin(radians)
    return new Matrix4(c, -s, 0, 0, s, c, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1)
  }
}
```

</details>


<details>
<summary>hack3d.c</summary>

```c
#include "xsmc.h"
#include "mc.xs.h"			// for xsID_ values
#include "xsHost.h"

void xs_Matrix(xsMachine *the)
{
	if ((1 == xsmcArgc) && xsmcTest(xsArg(0)))		// instanceof Float64Array and length === 16
		xsmcSet(xsThis, xsID_m, xsArg(0));
	else if (16 == xsmcArgc) {		// typeof first argument is number
		xsNumberValue d[16], *r;
		xsUnsignedValue rSize;
		xsmcVars(1);
		xsmcSetInteger(xsVar(0), 16);
		xsmcNew(xsVar(0), xsGlobal, xsID_Float64Array, &xsVar(0), NULL);
		xsmcSet(xsThis, xsID_m, xsVar(0));
		for (int i = 0; i < 16; i++)
			d[i] = xsmcToNumber(xsArg(i));
		
		xsmcGet(xsVar(0), xsVar(0), xsID_buffer);
		xsmcGetBufferWritable(xsVar(0), (void **)&r, &rSize);
		if (128 != rSize)
			xsUnknownError("unexpected");
		c_memmove(r, d, rSize);
	}
	else
		xsUnknownError("Invalid arguments");
}

void xs_matrix_multiply(xsMachine *the)
{
  xsNumberValue *m, *r, *result;
  xsUnsignedValue mSize, rSize, resultSize;
  xsSlot n;

  xsmcVars(3);
  xsmcSetInteger(n, 16);
  xsmcNew(xsVar(0), xsGlobal, xsID_Float64Array, &n, NULL);
  xsmcNew(xsResult, xsThis, xsID_constructor, &xsVar(0), NULL);

  xsmcGet(xsVar(0), xsThis, xsID_m);
  xsmcGet(xsVar(0), xsVar(0), xsID_buffer);

  xsmcGet(xsVar(1), xsArg(0), xsID_m);
  xsmcGet(xsVar(1), xsVar(1), xsID_buffer);

  xsmcGet(xsVar(2), xsResult, xsID_m);
  xsmcGet(xsVar(2), xsVar(2), xsID_buffer);

  xsmcGetBufferReadable(xsVar(0), (void **)&m, &mSize);
  xsmcGetBufferReadable(xsVar(1), (void **)&r, &rSize);
  xsmcGetBufferWritable(xsVar(2), (void **)&result, &resultSize);
  if ((mSize != 128) || (rSize != 128) || (resultSize != 128))
    xsUnknownError("invalid input");

  result[0] = m[0] * r[0] + m[1] * r[4] + m[2] * r[8] + m[3] * r[12];
  result[1] = m[0] * r[1] + m[1] * r[5] + m[2] * r[9] + m[3] * r[13];
  result[2] = m[0] * r[2] + m[1] * r[6] + m[2] * r[10] + m[3] * r[14];
  result[3] = m[0] * r[3] + m[1] * r[7] + m[2] * r[11] + m[3] * r[15];
  result[4] = m[4] * r[0] + m[5] * r[4] + m[6] * r[8] + m[7] * r[12];
  result[5] = m[4] * r[1] + m[5] * r[5] + m[6] * r[9] + m[7] * r[13];
  result[6] = m[4] * r[2] + m[5] * r[6] + m[6] * r[10] + m[7] * r[14];
  result[7] = m[4] * r[3] + m[5] * r[7] + m[6] * r[11] + m[7] * r[15];
  result[8] = m[8] * r[0] + m[9] * r[4] + m[10] * r[8] + m[11] * r[12];
  result[9] = m[8] * r[1] + m[9] * r[5] + m[10] * r[9] + m[11] * r[13];
  result[10] = m[8] * r[2] + m[9] * r[6] + m[10] * r[10] + m[11] * r[14];
  result[11] = m[8] * r[3] + m[9] * r[7] + m[10] * r[11] + m[11] * r[15];
  result[12] = m[12] * r[0] + m[13] * r[4] + m[14] * r[8] + m[15] * r[12];
  result[13] = m[12] * r[1] + m[13] * r[5] + m[14] * r[9] + m[15] * r[13];
  result[14] = m[12] * r[2] + m[13] * r[6] + m[14] * r[10] + m[15] * r[14];
  result[15] = m[12] * r[3] + m[13] * r[7] + m[14] * r[11] + m[15] * r[15];
}

void xs_matrix_multiplyVec4(xsMachine *the)
{
  xsNumberValue wIn, xIn, yIn, zIn;
  xsNumberValue wOut, xOut, yOut, zOut;
  xsNumberValue *m;
  xsUnsignedValue mSize;
  xsSlot n;

  xsmcVars(1);

  xsmcGet(n, xsArg(0), xsID_w); wIn = xsmcToNumber(n);
  xsmcGet(n, xsArg(0), xsID_x); xIn = xsmcToNumber(n);
  xsmcGet(n, xsArg(0), xsID_y); yIn = xsmcToNumber(n);
  xsmcGet(n, xsArg(0), xsID_z); zIn = xsmcToNumber(n);

  xsmcGet(xsVar(0), xsThis, xsID_m);
  xsmcGet(xsVar(0), xsVar(0), xsID_buffer);
  xsmcGetBufferReadable(xsVar(0), (void **)&m, &mSize);
  if (128 != mSize)
    xsUnknownError("invalid");

  xOut = m[0] * xIn + m[1] * yIn + m[2] * zIn + m[3] * wIn;
  yOut = m[4] * xIn + m[5] * yIn + m[6] * zIn + m[7] * wIn;
  zOut = m[8] * xIn + m[9] * yIn + m[10] * zIn + m[11] * wIn;
  wOut = m[12] * xIn + m[13] * yIn + m[14] * zIn + m[15] * wIn;

  xsmcSetNewObject(xsResult);
  xsmcSetNumber(n, xOut); xsmcSet(xsResult, xsID_x, n);
  xsmcSetNumber(n, yOut); xsmcSet(xsResult, xsID_y, n);
  xsmcSetNumber(n, zOut); xsmcSet(xsResult, xsID_z, n);
  xsmcSetNumber(n, wOut); xsmcSet(xsResult, xsID_w, n);
}
```

</details>

I also tweaked `calculateBounds` to reduce some overhead, since it was high in the profile. The code was fine for most situations, but this is a performance hotspot.

```js
function calcBounds(vertices: Vec4[]): Bounds {
  let x1 = vertices[0].x, y1 = vertices[0].y, x2 = vertices[0].x, y2 = vertices[0].y;
  for (let i = 0, length = vertices.length; i < length; i++) {
    const v = vertices[i];
    // +/-1 is due to width of stroke in Outline.stroke()
    if (v.x <= x1) x1 = v.x - 1
    if (v.x >= x2) x2 = v.x + 1
    if (v.y <= y1) y1 = v.y - 1
    if (v.y >= y2) y2 = v.y + 1
  }
  return {x1, x2, y1, y2}
}
```

**Reply by phoddie** [reply]
@tve – curious if you had a chance to give this a try? If you prefer, I could make a PR to your repo.

**Reply by tve** [reply]
A PR is definitely appreciated, but I'm scrambling for a project (that uses the IMU but not the 3D viz) so this is going to be stuck in the queue for a few more days...

**Reply by phoddie** [reply]
PR is now there. https://github.com/tve/modtve/pull/1

FWIW – one more change is needed to make this build without changes: the fusion library would either need to be included in modtve/lib/fusion or added to the manifest as a Git import. I wasn't sure about your preference so I just left it as-is.
